### ISSUE TITLE
Failed to fetch http://nvidia.github.io/nvidia-docker/ubuntu18.04/amd64/InRelease

### ISSUE THREAD

**Issue by weberhen** [NVIDIA/nvidia-docker#1126]
### 1. Issue or feature description

apt-get update and apt-get install nvidia-container-toolkit fails on ubuntu 18.04

### 2. Steps to reproduce the issue

$ distribution=$(. /etc/os-release;echo $ID$VERSION_ID)
$ curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | sudo apt-key add -
$ curl -s -L https://nvidia.github.io/nvidia-docker/$distribution/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list
$ sudo apt-get update && sudo apt-get install -y nvidia-container-toolkit

output of sudo apt-get update contains -

Hit:1 http://archive.canonical.com bionic InRelease
Err:2 https://nvidia.github.io/libnvidia-container/ubuntu18.04/amd64  InRelease                                                   
  Connection failed [IP: 185.199.109.153 443]
Hit:5 http://archive.ubuntu.com/ubuntu bionic InRelease                   
Err:3 https://nvidia.github.io/nvidia-container-runtime/ubuntu18.04/amd64  InRelease                                             
  Connection failed [IP: 185.199.108.153 443]
Err:4 https://nvidia.github.io/nvidia-docker/ubuntu18.04/amd64  InRelease                                                        
  Connection failed [IP: 185.199.109.153 443]
Get:6 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB] 
Get:7 http://archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]     
Get:8 http://archive.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]      
Fetched 252 kB in 2s (141 kB/s)                                                 
Reading package lists... Done
W: Failed to fetch http://nvidia.github.io/libnvidia-container/ubuntu18.04/amd64/InRelease  Connection failed [IP: 185.199.109.153 443]
W: Failed to fetch http://nvidia.github.io/nvidia-container-runtime/ubuntu18.04/amd64/InRelease  Connection failed [IP: 185.199.108.153 443]
W: Failed to fetch http://nvidia.github.io/nvidia-docker/ubuntu18.04/amd64/InRelease  Connection failed [IP: 185.199.109.153 443]
W: Some index files failed to download. They have been ignored, or old ones used instead.

### 3. Information to [attach](https://help.github.com/articles/file-attachments-on-issues-and-pull-requests/) (optional if deemed irrelevant)

 - [ ] Some nvidia-container information: `nvidia-container-cli -k -d /dev/tty info`
 - [x] Kernel version from `uname -a`
`Linux STATION-E8068 5.0.0-32-generic #34~18.04.2-Ubuntu SMP Thu Oct 10 10:36:02 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`
 - [ ] Any relevant kernel output lines from `dmesg`
 - [x] Driver information from `nvidia-smi -a`
| NVIDIA-SMI 435.21       Driver Version: 435.21       CUDA Version: 10.1     |
|   0  GeForce RTX 2060    Off  | 00000000:01:00.0  On |                  N/A |
  - [x] Docker version from `docker version`
Client: Docker Engine - Community
 Version:           19.03.2
 API version:       1.40
 Go version:        go1.12.8
 Git commit:        6a30dfc
 Built:             Thu Aug 29 05:29:11 2019
 OS/Arch:           linux/amd64
 Experimental:      false
 Server: Docker Engine - Community
 Engine:
  Version:          19.03.2
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.8
  Git commit:       6a30dfc
  Built:            Thu Aug 29 05:27:45 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

 - [ ] NVIDIA packages version from `dpkg -l '*nvidia*'` _or_ `rpm -qa '*nvidia*'`
 - [ ] NVIDIA container library version from `nvidia-container-cli -V`
 - [ ] NVIDIA container library logs (see [troubleshooting](https://github.com/NVIDIA/nvidia-docker/wiki/Troubleshooting))
 - [ ] Docker command, image and tag used

**Comment by weberhen** [comment]
I tried the solution provided in #1064 but it does not work.

**Comment by lsimmons2** [comment]
I'm getting a similar issue. After installing CUDA drivers/toolkit on Ubuntu 18.04.03 with instructions [here](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#ubuntu-installation) I'm following the Quickstart instructions in this repo's README and getting the error below:

    # Add the package repositories
    $ distribution=$(. /etc/os-release;echo $ID$VERSION_ID)
    $ curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | sudo apt-key add -
    $ curl -s -L https://nvidia.github.io/nvidia-docker/$distribution/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list

    $ sudo apt-get update

    >>>

    Get:1 file:/var/cuda-repo-10-1-local-10.1.243-418.87.00  InRelease
    Ign:1 file:/var/cuda-repo-10-1-local-10.1.243-418.87.00  InRelease
    Get:2 file:/var/cuda-repo-10-1-local-10.1.243-418.87.00  Release [574 B]
    Get:2 file:/var/cuda-repo-10-1-local-10.1.243-418.87.00  Release [574 B]
    Hit:4 https://nvidia.github.io/libnvidia-container/ubuntu18.04/amd64  InRelease                                  
    Hit:5 https://nvidia.github.io/nvidia-container-runtime/ubuntu18.04/amd64  InRelease                             
    Hit:6 https://nvidia.github.io/nvidia-docker/ubuntu18.04/amd64  InRelease                                        
    Hit:7 https://dl.yarnpkg.com/debian stable InRelease                                                             
    Ign:8 http://dl.google.com/linux/chrome/deb stable InRelease                                                     
    Hit:10 http://us.archive.ubuntu.com/ubuntu bionic InRelease                                                      
    Hit:11 http://dl.google.com/linux/chrome/deb stable Release                                                      
    Err:12 https://get.docker.com/ubuntu docker InRelease                                                            
      403  Forbidden [IP: 2600:9000:20a6:5600:10:a463:3d00:93a1 443]
    Hit:13 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease                                     
    Hit:14 https://download.docker.com/linux/ubuntu bionic InRelease                                                 
    Get:15 http://us.archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]                                    
    Get:16 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]                                
    Hit:9 https://packagecloud.io/slacktechnologies/slack/debian jessie InRelease                                    
    Get:17 http://us.archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]                                  
    Reading package lists... Done                                                                     
    E: Failed to fetch https://get.docker.com/ubuntu/dists/docker/InRelease  403  Forbidden [IP: 2600:9000:20a6:5600:10:a463:3d00:93a1 443]
    E: The repository 'https://get.docker.com/ubuntu docker InRelease' is not signed.
    N: Updating from such a repository can't be done securely, and is therefore disabled by default.
    N: See apt-secure(8) manpage for repository creation and user configuration details.

Checking `/etc/apt/sources.list.d/nvidia-docker.list` the file contains:

    deb https://nvidia.github.io/libnvidia-container/ubuntu18.04/$(ARCH) /
    deb https://nvidia.github.io/nvidia-container-runtime/ubuntu18.04/$(ARCH) /
    deb https://nvidia.github.io/nvidia-docker/ubuntu18.04/$(ARCH) /

Is this correct? Should $ARCH have expanded? Manually curling `https://nvidia.github.io/libnvidia-container/ubuntu18.04/x86_64` (and the other two equivalent URLs) gives me 404s with the following message:

    # Unsupported distribution!
    # Check https://nvidia.github.io/libnvidia-container

Other potentially relevant info:
    $ nvidia-smi -a | grep version
    >>> Driver Version                      : 418.87.00

    $ gcc --version
    >>> gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0

    $ uname -a
    >>> Linux <my_comp_name> 5.0.0-36-generic #39~18.04.1-Ubuntu SMP Tue Nov 12 11:09:50 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

    $ docker --version
    >>> Docker version 19.03.4, build 9013bf583a

**Comment by nvjmayo** [comment]
> Is this correct? Should $ARCH have expanded? Manually curling https://nvidia.github.io/libnvidia-container/ubuntu18.04/x86_64 (and the other two equivalent URLs) gives me 404s with the following message:

FYI there is no directory listing on the website, so you'll get a 404 unless you put in an exact URL. Example: 
`curl --head https://nvidia.github.io/libnvidia-container/ubuntu18.04/amd64/Packages.xz`

And the other thing to note is that the architecture used in the URL is **amd64** not _x86_64_.

**Comment by weberhen** [comment]
I have the same content as @lsimmons2 in /etc/apt/sources.list.d/nvidia-docker.list, and even if I manually change it to 

deb https://nvidia.github.io/libnvidia-container/ubuntu18.04/amd64 /
deb https://nvidia.github.io/nvidia-container-runtime/ubuntu18.04/amd64 /
deb https://nvidia.github.io/nvidia-docker/ubuntu18.04/amd64 /

I still get the same errors as before.

**Comment by lsimmons2** [comment]
@nvjmayo thanks for the tips. So curling the correct URLs returns 200s but I take it those URLs are not what should be in `/etc/apt/sources.list.d/nvidia-docker.list`?

I've tried the following in `/etc/apt/sources.list.d/nvidia-docker.list`:
- `deb https://nvidia.github.io/libnvidia-container/ubuntu18.04/ bionic main`
- `deb https://nvidia.github.io/libnvidia-container/ubuntu18.04/amd64 bionic main`
-`deb https://nvidia.github.io/libnvidia-container/ubuntu18.04/amd64/Packages.xz bionic main`

and they all produce the following error (with respective URLs) with `apt-get update`:
```
Err:9 https://nvidia.github.io/libnvidia-container/ubuntu18.04 bionic Release                                                                                                                                                        
  404  Not Found [IP: 185.199.111.153 443]
```

**Comment by weberhen** [comment]
I tried on another computer (ubuntu 16.04) and the contents of /etc/apt/sources.list.d/nvidia-docker.list were:

deb https://nvidia.github.io/libnvidia-container/ubuntu16.04/$(ARCH) /
deb https://nvidia.github.io/nvidia-container-runtime/ubuntu16.04/$(ARCH) /
deb https://nvidia.github.io/nvidia-docker/ubuntu16.04/$(ARCH) /

In that computer, the installation was successful.

**Comment by RenaudWasTaken** [comment]
> @nvjmayo thanks for the tips. So curling the correct URLs returns 200s but I take it those URLs are not what should be in /etc/apt/sources.list.d/nvidia-docker.list?

@lsimmons2 on ubuntu, the ARCH variable should resolve to x86_64 not amd64.

> Linux STATION-E8068 5.0.0-32-generic #34~18.04.2-Ubuntu SMP Thu Oct 10 10:36:02 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

Funnily enough your `uname -m` seems to resolve to the correct definition. I'm guessing `dpkg --printarchitecture` resolves to `amd64`, you could replace the `$(ARCH)` variable in the source.list by `x86_64` as a workaround.
Note that you should probably investigate why dpkg is not using the system's architecture.

**Comment by weberhen** [comment]
solved my problem in 18.04 by installing manually all .deb packages (wget http:package.deb, then dpkg -i package.deb)

**Comment by lhl2017** [comment]
I resolved this issue to use amd64 in place of  $(ARCH).
`  1 deb https://nvidia.github.io/libnvidia-container/ubuntu18.04/amd64 /
  2 deb https://nvidia.github.io/nvidia-container-runtime/ubuntu18.04/amd64 /
  3 deb https://nvidia.github.io/nvidia-docker/ubuntu18.04/amd64 /`
